rfd/0165-ssh-connection-handover.md

```diff
@@ -0,0 +1,52 @@
+
+[rfd0150]: https://github.com/gravitational/teleport/blob/master/rfd/0150-ssh-connection-resumption.md
+
+### Connection registry
+
+Each new resumption-enabled connection (with a given _resumption token_) will be associated with a listener in the hand-over connection registry, implemented, on UNIX platforms, as the directory `<datadir>/handover/`, where each connection is associated with a UNIX stream listener at `<datadir>/handover/<SHA-256 hash of resumption token>.sock`; the socket should be unlinked when the resumable connection is closed, and each Teleport process will, during startup, clean up sockets potentially left over by Teleport processes that were terminated before they could clean up after themselves.
+
+The `handover` directory will be created with permissions such that only the user that created the directory (i.e. the user under which the Teleport binary is running) is allowed to connect to sockets in it. Sockets should not be connectable by anyone except the user running Teleport as the caller will be responsible for propagating the IP address of the client, and thus should be trusted.
+
+The name of the socket should be a hash of the resumption token, to harden this mechanism against potential local information leaks. It's sufficient to use a regular hash function (rather than something keyed by a secret like a HMAC) because the resumption token is a 128-bit value picked uniformly, and thus unfeasible to bruteforce directly.
+
+### Hand-over protocol
+
+After reading the resumption token from the client, the server will first check if the token is associated with a local connection. If found, connection resumption proceeds as usual. If not, the server will attempt to connect to the hand-over listener for that token. If successful, it will send the client's IP address over the connection (as a 16 byte IPv6 address, mapping the IPv4 to IPv6 if needed), then it will start bidirectionally copying data between the client and the hand-over connection. If the connection to the hand-over listener is not successful because the socket doesn't exist, then the server will reply with a "connection not found" tag to the client.
+
+The server-side of the hand-over connection will first read the remote IP address, then proceed operating with the connection as if a client with the received remote IP address had just attempted to resume a connection with the token associated to the hand-over connection, by replying with a "success" tag if the remote address matches the one of the resumable connection, a "bad address" tag otherwise, or a "not found" tag if the connection had been closed in the meantime.
```

Comment:
It's assumed that whoever can connect to the handover sockets can be trusted to provide correct client IP information; thus the reason why the sockets are only handled in a directory with restrictive perms (because permissions on unix sockets can be somewhat weird on certain BSDs, but all directories in the path to the socket are always checked for read and access).

Reply:
I'll update the wording to clarify, but the client IP information is provided by the other Teleport process, not by the remote client in any way - and the other Teleport process gets the client IP info from the proxy through the (authenticated) reverse tunnel, as usual.